File: 2023/leetcode/subarray_sums_divisible_by_k_leetcode_974.py
from typing import List


def getPrefixArray(nums):
    prefix_array = [0 for _ in range(len(nums))]
    total = 0
    for x in range(0, len(nums)):
        total += nums[x]
        prefix_array[x] = total
    return prefix_array


class Solution:
    def subarraysDivByK(self, nums: List[int], k: int) -> int:
        prefix_sum = getPrefixArray(nums)
        count = 0
        for x in prefix_sum:
            if (x % k == 0):
                count += 1
        
        print(prefix_sum)

        c = 1
        while(len(prefix_sum) > 0):
            if (prefix_sum[c] % k == 0):
                count += 1

            c+=1
            prefix_sum.pop(0)


# ans = Solution().subarraysDivByK(nums=[1, -3, 2, 3], k=2)
ans = Solution().subarraysDivByK(nums = [4,5,0,-2,-3,1], k = 5)
print(ans)

File: leetcode/subarray_sums_divisible_by_k_leetcode_974.cpp
#include <bits/stdc++.h>
using namespace std;

void printVector(string tag, vector<int> arr)
{
	cout << tag << " --> ";
	for (int i = 0; i < arr.size(); i++)
	{
		cout << arr[i] << " ";
	}
	cout << endl;
	// cout << "capacity: " << arr.capacity() << endl;
	cout << tag << " size: " << arr.size() << endl;
}

void printUMAP(unordered_map<int, int> &umap){
	 // Traversing an unordered map
  for (auto x : umap)
    cout << x.first << " " << 
            x.second << endl;
}

class Solution
{
public:
	int subarraysDivByK(vector<int> &nums, int k)
	{

		unordered_map<int, int> mp;

		// we have seen 0 atleast 1 times
		mp[0] = 1;

		int total = 0;
		int answer = 0;
		for (int i=0; i < nums.size(); i++){
			total += nums[i];

			int rem = total % k;

			if (mp[i] < 0){
				rem += k;
			}

			// check if rem is seen previously in map
			if (mp.find(rem) != mp.end()){
				answer += mp[rem];
			}

			mp[rem]++;
		}  

		printVector("nums -->", nums);
		printUMAP(mp);
		cout << "ans -->  " << answer << endl;

		return -1;
	}
};

int main()
{
	vector<int> nums = {4, 5, 0, -2, -3, 1};
	Solution sol;
	sol.subarraysDivByK(nums, 5);
}